### ISSUE TITLE
CTR encryption with pre-defined nonce and counter

### ISSUE THREAD

**Issue by JohnDingeling** [Legrandin/pycryptodome#399]
CTR encryption with pre-defined nonce and counter

I have to write a python script, which shall reproduce a AES CTR encryption from an embedded device i.e. the python script shall produce the exact same cipher text for the same given plain text, same key and initialization vector (IV, consisting of nonce and counter). However, I am not able to achieve this with the pycryptodome library and start asking myself, whether this particular use case can be done at all with pycryptodome. I would really appreciate if you could help me by either confirming that this cannot be done with pycryptodome or (even better), tell how ot use the pycryptodome API in a way, that my use case works out.

I have certain restrictions regarding the IV which are coming from the embedded device (block size is 128 bit/16 bytes, thus key and IV have 128 bit length):
- the nonce part of the IV has to consist of 12 byte, 8 bytes random number + 4 bytes fixed zeros, e.g. 0x08,0x07,0x06,0x05,0x04,0x03,0x02,0x01,0x00,0x00,0x00,0x00
- the counter part of the IV has to start with following four bytes: 0x00,0x00,0x00,0x09
So the IV as it is used on the embedded device looks like this as a byte stream: 0x08,0x07,0x06,0x05,0x04,0x03,0x02,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x09.
The embedded device uses little-endian storage of bytes in memory. What I can monitor on the embedded device is, that the IV, i.e. the counter part of it, is incremented after each encrypted AES block (=16 bytes) by 1, so the IV evolves for each block as follows:
08070605040302010000000001000009
08070605040302010000000002000009
08070605040302010000000003000009
08070605040302010000000004000009
08070605040302010000000005000009
08070605040302010000000006000009
0807060504030201000000000700000
...
For simplicity while testing  I set the AES key just to 16 bytes of zeros (0x00).
I did set up the pycryptodome objects as follows:

aes_nonce = bytearray([0x08,0x07,0x06,0x05,0x04,0x03,0x02,0x01,0x00,0x00,0x00,0x00])
myIv = Crypto.Util.Counter.new(32, prefix=aes_nonce, little_endian=True, initial_value=0x09000000) #12 bytes of nonce + 4 bytes of counter

aes_key = bytearray([0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]) #16 bytes of 0x00 just for testing purposes
plaintext = bytearray([0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0a,0x0b,0x0c,0x0d,0x0e,0x0f,0x10])
encryptor = Crypto.Cipher.AES.new(aesKeyWordSwapped, AES.MODE_CTR, counter=myIv1)
cipher_text = encryptor.encrypt(plaintext)

The problem though is, that I always get a different cipher text on the embedded device and in the python script for the same input, even in my tests with only one 16 bytes block of given input plain text. I tried all sorts of endianess variations for the input parameters of the Crypto.Util.Counter object but it wouldn' help. It is quite hard to debug as there is no way to see how the nonce and counter part of the IV are put together and how exactly the counter is incremented internally in the pycryptodome library.

With the plain text, IV and key set up as defined above I get following cipher texts:
Embedded device:	017f2178f8381c08d576d89d9061f96e
Python script:		0a7a8dd96086e5d5ce46aba9db526703
					
Any hints or help would be very much appreciated!

**Comment by texadactyl** [comment]
I was able to verify your Python script.  I suspect that there is something wrong on your embedded device.  See my somewhat modified attachment.
[ctr_crypto_enc_dec.py.txt](https://github.com/Legrandin/pycryptodome/files/4415038/ctr_crypto_enc_dec.py.txt)

Sample output:
```
Nonce
0000   08 07 06 05 04 03 02 01 00 00 00 00               ............     

IV: {'counter_len': 4, 'prefix': bytearray(b'\x08\x07\x06\x05\x04\x03\x02\x01\x00\x00\x00\x00'), 'suffix': b'', 'initial_value': 150994944, 'little_endian': True}
AES key
0000   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   ................ 

Cleartext, raw
0000   01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10   ................ 

Ciphertext
0000   0a 7a 8d d9 60 86 e5 d5 ce 46 ab a9 db 52 67 03   .z..`....F...Rg. 

Cleartext, after decrypting
0000   01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10   ................ 
```

**Comment by texadactyl** [comment]
Also, you can do this instead of a zero key:
```
from Crypto.Random import get_random_bytes
aes_key = get_random_bytes(16)
```

Sample output:
```
Nonce
0000   08 07 06 05 04 03 02 01 00 00 00 00               ............     

IV: {'counter_len': 4, 'prefix': bytearray(b'\x08\x07\x06\x05\x04\x03\x02\x01\x00\x00\x00\x00'), 'suffix': b'', 'initial_value': 150994944, 'little_endian': True}
AES key
0000   6f 6d 5d 77 3e a3 ee c2 a6 a8 7d 10 da 31 fe ba   om]w>.....}..1.. 

Cleartext, raw
0000   01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10   ................ 

Ciphertext
0000   ff a5 7c 2c 72 e5 2d dc a3 d0 b0 16 48 ad 3e e7   ..|,r.-.....H.>. 

Cleartext, after decrypting
0000   01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10   ................ 
```

**Comment by JohnDingeling** [comment]
Hi @texadactyl, thanks for your feedback. So it looks that my assumption is correct, that the Counter object builds the IV just by doing a merge of prefix (nonce) + initial_value (counter), this also matches the documentation. Meanwhile I could also verify, that another Python library (cryptography) yields the same cipher text (0a7a8dd96086e5d5ce46aba9db526703) for the same input.
So it must be a peculiarity of the embedded devices AES CTR encryption. The thing is, the embedded device is also able to decrypt its own cipher text (017f2178f8381c08d576d89d9061f96e) correctly. So guess I'll have to continue digging...

**Comment by texadactyl** [comment]
@JohnDingeling 
I take it back; there is probably nothing wrong on the embedded device.  I believe that what you have is 2 different environments (Python on PC versus ? on the ? embedded device), each consistent within their own framework, but doing something different with the IV set up with respect to each other.

For example, I have a Java program which I had hoped would be equivalent to the Python program (ctr_crypto_enc_dec.py).  Guess what happened to me?  Yes, I am getting 2 different ciphertext displays yet both have no trouble being encrypt/decrypt within themselves.

So, I changed the Python program from doing this:
```
ctr_vector = Counter.new(32, #bits of counter (4 bytes))
                   prefix=aes_nonce,
                   little_endian=True,
                   initial_value=0x09000000) #12 bytes of nonce + 4 bytes of counter
encryptor = AES.new(aes_key, AES.MODE_CTR, counter=ctr_vector)
```
to do this instead:
```
encryptor = AES.new(aes_key, AES.MODE_CTR, nonce=aes_nonce, initial_value=0x09000000)
```
Now, the Python program and the Java program agree but on a different ciphertext value.
New Python: ```fa b6 2c 6b c2 1d d9 82 c2 be 31 53 4d 20 6f db```
Original Python: ```0a 7a 8d d9 60 86 e5 d5 ce 46 ab a9 db 52 67 03```
"embedded device": ```01 7f 21 78 f8 38 1c 08 d5 76 d8 9d 90 61 f9 6e```

[TryCTR.java.txt](https://github.com/Legrandin/pycryptodome/files/4422682/TryCTR.java.txt)
[TryCTR.py.txt](https://github.com/Legrandin/pycryptodome/files/4422683/TryCTR.py.txt)

(continued)

**Comment by texadactyl** [comment]
So, I went back to the original method but changed to Big Endian:
```
ctr_vector = Counter.new(32, #bits of counter (4 bytes))
                   prefix=aes_nonce,
                   little_endian=False,
                   initial_value=0x09000000) #12 bytes of nonce + 4 bytes of counter
encryptor = AES.new(aes_key, AES.MODE_CTR, counter=ctr_vector)
```

The result is that now the Big Endian form of the original Python program agrees with the Java program but not with the embedded device.  If I wanted to get the Java program to use Little Endian form of the original Python program, I'd have to figure out how to stop the Java JCE from defaulting to Big Endian processing of the counter subfield of the IV.  Ugh.

I would have expected your embedded device to agree with Big Endian if it was also using MODE_CTR.

So, my questions:

1. What is this embedded device?  Android?  iOS?
2. What language are you programming in?  Python?  Java?
3. Which cryptographic toolkit are you using?
4. Are you using a crypto "provider" (E.g. Bouncy Castle") or just defaulting to the toolkit?

**Comment by texadactyl** [comment]
Found out that Oracle JCE for Java follows the NIST guidelines which do not allow "Little Endian".  See appendix B in https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf

So, while not a happy ending, my observations and the Oracle JCE documentation are consistent.

**Comment by texadactyl** [comment]
@JohnDingeling 
You owe my a pint of Old Peculiar when this COVID-19 tragedy is over.  (=:

**Comment by JohnDingeling** [comment]
Hi @texadactyl , sorry for my late reply....I meanwhile found out the solution: My embedded device uses a STM32 chip with hw peripherals for AES encryption with really strange habits of flipping bytes around for optimized memory access or what not. However, in the python script I had to word-wise (i.e. for each 4 byte group) swap the byte order, for the plaintext, the IV, the key and also in the and for the cipher text. This way I was able to produce the same cipher text as the embedded device with the python script.
Sure, will spend you a pint of whatever for your efforts ;-)
